### PR TITLE
Add API Authentication

### LINKED ISSUE #1: Authentication implementation
Using Laravel Sanctum & Fortify

### PR DESCRIPTION
**Pull request by calumshep**
Closes #1 

- Adds authentication as implementation of OAuth 2 using Laravel Passport
- Adds new `php artisan data:regenerate` command to refresh application test data
- Adds new controller to manage users including administrative tasks and profile editing
- Adds middleware for checking user role for user models

### CODE REVIEW

File: app/Http/Controllers/AuthController.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function register(Request $request)
+    {
+        $validator = Validator::make($request->all(), [
+            'username' => 'required|min:2|max:55|unique:users',
+            'name' => 'required|min:2|max:255',
+            'email' => 'email|required|unique:users',
+            'password' => 'required|confirmed|min:6',
+            'date_of_birth' => 'required|date'
```
Comment (wrt95) [inline app/Http/Controllers/AuthController.php:27]: Can we add a minimum age of for example 13?

File: app/Http/Controllers/AuthController.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     * Register a new user
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function register(Request $request)
+    {
+        $validator = Validator::make($request->all(), [
+            'username' => 'required|min:2|max:55|unique:users',
```
Comment (wrt95) [inline app/Http/Controllers/AuthController.php:23]: can we make max length smaller? 55 will be a lot to display front end. For example 25?

File: app/Http/Controllers/UserController.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+     *
+     * @param \Illuminate\Http\Request $request
+     * @param \App\Models\User $user
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function update(Request $request, User $user)
+    {
+        $validator = Validator::make($request->all(), [
+            'username'  => 'required|max:55',
```
Comment (wrt95) [inline app/Http/Controllers/UserController.php:53]: can we make max length smaller? 55 will be a lot to display front end. For example 25?

**Review by wrt95 — CHANGES_REQUESTED** [review]
Overall looks good, but I feel we need age validation.

File: app/Http/Controllers/AuthController.php (hunk 2 above)
Comment (calumshep) [reply app/Http/Controllers/AuthController.php:23]: Will reduce the username to 20 and reduce the max name to...30?

File: app/Http/Controllers/UserController.php (hunk 3 above)
Comment (calumshep) [reply app/Http/Controllers/UserController.php:53]: See https://github.com/Watt-Dev/snapthat-api/pull/3#discussion_r576721047

**Review by wrt95 — DISMISSED** [review]

**Review by wrt95 — APPROVED** [review]

**Pull request merged**